File: heap/lastStoneWeight.cpp
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int lastStoneWeight(vector<int>& stones) {
    priority_queue<int> maxHeap(stones.begin(), stones.end());

    while (maxHeap.size() > 1) {
        // Fetch top two stones:
        const int heavierStone = maxHeap.top(); maxHeap.pop();
        const int lighterStone = maxHeap.top(); maxHeap.pop();

        const int newStone = heavierStone - lighterStone;
        if (newStone) maxHeap.push(newStone);
    }

    return maxHeap.empty() ? 0 : maxHeap.top();
}

// Test cases:
int main() {
    vector<int> stones{2, 7, 4, 1, 8, 1};
    cout << lastStoneWeight(stones) << endl; // Output: 1

    vector<int> stones2{1, 3};
    cout << lastStoneWeight(stones2); // Output: 2
    cout << endl;

    return 0;
}

File: heap/last_stone_weight.py
import heapq


def last_stone_weight(stones: list[int]) -> int:
    max_heap = []
    for stone in stones:
        heapq.heappush(max_heap, -stone)
    
    while len(max_heap) > 1:
        heavy_stone_1 = -heapq.heappop(max_heap)
        heavy_stone_2 = -heapq.heappop(max_heap)
        
        new_stone = heavy_stone_1 - heavy_stone_2
        heapq.heappush(max_heap, -new_stone)
        
    return 0 if len(max_heap) == 0 else -max_heap[0]

# Test cases:
# Test case 1:
print(last_stone_weight([2, 7, 4, 1, 8, 1]))  # Output: 1
# Test case 2:
print(last_stone_weight([1, 3]))  # Output: 2